File: app/src/main/java/org/ametro/utils/AnimationInterpolator.java
package org.ametro.utils;

import android.graphics.PointF;

public class AnimationInterpolator {

	public final static int SCALE = 1;
	public final static int SCROLL = 2;
	public final static int SCALE_AND_SCROLL = 3;
	
	private PointF mStartPoint = new PointF();
	private PointF mEndPoint = new PointF();
	private PointF mNowPoint = new PointF();
	
	private float mStartScale;
	private float mEndScale;
	
	private long mPeriod;
	private long mEndTime;
	private long mNowTime;
	
	private int mMode;
	
	public void begin(PointF startPoint, PointF endPoint, Float startScale, Float endScale, long time){
		this.mNowTime = System.currentTimeMillis();
		this.mEndTime = mNowTime + time;
		this.mPeriod = time;
		this.mMode = 0;
		if(startScale!=null && endScale!=null){
			this.mStartScale = startScale;
			this.mEndScale = endScale;
			this.mMode |= SCALE;
		}
		if(startPoint!=null && endPoint!=null){
			this.mStartPoint.set(startPoint);
			this.mEndPoint.set(endPoint);
			this.mNowPoint.set(startPoint);
			this.mMode |= SCROLL;
		}
	}

    public boolean more(){
		return mNowTime < mEndTime; 
	}
	
	public void next(){
		mNowTime = System.currentTimeMillis();
	}
	
	public PointF getPoint(){
		if(mNowTime < mEndTime){
			float k = getProgress();
			float x = mEndPoint.x - k * (mEndPoint.x - mStartPoint.x);
			float y = mEndPoint.y - k * (mEndPoint.y - mStartPoint.y);
			mNowPoint.set(x, y);
			return mNowPoint;
		}else{
			return mEndPoint;
		}
	}
	
	public float getScale(){
		if(mNowTime < mEndTime){
			float k = getProgress();
			return (mEndScale - k * (mEndScale - mStartScale));
		}else{
			return mEndScale;
		}
	}
	
	private float getProgress() {
		return (float)(mEndTime - mNowTime) / mPeriod;
	}

	public boolean hasScale() {
		return (mMode & SCALE)!=0;
	}
	
	public boolean hasScroll() {
		return (mMode & SCROLL)!=0;
	}
	
}

File: app/src/main/java/io/github/romangolovanov/apps/ametro/utils/AnimationInterpolator.kt
package io.github.romangolovanov.apps.ametro.utils

import android.graphics.PointF

class AnimationInterpolator {

    private val mStartPoint = PointF()
    private val mEndPoint = PointF()
    private val mNowPoint = PointF()

    private var mStartScale = 0f
    private var mEndScale = 0f

    private var mPeriod = 0L
    private var mEndTime = 0L
    private var mNowTime = 0L

    private var mMode = 0

    fun begin(startPoint: PointF?, endPoint: PointF?, startScale: Float?, endScale: Float?, time: Long) {
        mNowTime = System.currentTimeMillis()
        mEndTime = mNowTime + time
        mPeriod = time
        mMode = 0
        if (startScale != null && endScale != null) {
            mStartScale = startScale
            mEndScale = endScale
            mMode = mMode or SCALE
        }
        if (startPoint != null && endPoint != null) {
            mStartPoint.set(startPoint)
            mEndPoint.set(endPoint)
            mNowPoint.set(startPoint)
            mMode = mMode or SCROLL
        }
    }

    fun more(): Boolean = mNowTime < mEndTime

    fun next() {
        mNowTime = System.currentTimeMillis()
    }

    fun getPoint(): PointF {
        return if (mNowTime < mEndTime) {
            val k = getProgress()
            val x = mEndPoint.x - k * (mEndPoint.x - mStartPoint.x)
            val y = mEndPoint.y - k * (mEndPoint.y - mStartPoint.y)
            mNowPoint.set(x, y)
            mNowPoint
        } else {
            mEndPoint
        }
    }

    fun getScale(): Float {
        return if (mNowTime < mEndTime) {
            val k = getProgress()
            mEndScale - k * (mEndScale - mStartScale)
        } else {
            mEndScale
        }
    }

    private fun getProgress(): Float = (mEndTime - mNowTime).toFloat() / mPeriod

    fun hasScale(): Boolean = (mMode and SCALE) != 0

    fun hasScroll(): Boolean = (mMode and SCROLL) != 0

    companion object {
        const val SCALE = 1
        const val SCROLL = 2
        const val SCALE_AND_SCROLL = 3
    }
}
